### PR TITLE
[lldb] Encode operands and arity in Dwarf.def and use them in LLDB.

### PR DESCRIPTION
**Pull request by JDevlieghere**
This PR extends Dwarf.def to include the number of operands and the arity (the number of entries on the DWARF stack) and use it from the LLDB DWARF expression evaluator.

### CODE REVIEW

**Comment by JDevlieghere** [comment]
The number of operands comes directly from Table 7.6 in the DWARF 5 spec. I'm less confident in the _arity_ which required me to consult the description of the operations. I think I got them right and everything covered by the LLDB test suite still passes, but I would definitely appreciate a second pair of eyes on this.

File: llvm/include/llvm/BinaryFormat/Dwarf.def [hunk 1]
```diff
@@ -694,197 +694,197 @@ HANDLE_DW_FORM(0x1f21, GNU_strp_alt, 0, GNU)
...
+HANDLE_DW_OP(0x26, shra, 0, 2, 2, DWARF)
+HANDLE_DW_OP(0x27, xor, 0, 2, 2, DWARF)
+HANDLE_DW_OP(0x28, bra, 1, 1, 2, DWARF)
+HANDLE_DW_OP(0x29, eq, 0, 2, 2, DWARF)
+HANDLE_DW_OP(0x2a, ge, 0, 2, 2, DWARF)
+HANDLE_DW_OP(0x2b, gt, 0, 2, 2, DWARF)
+HANDLE_DW_OP(0x2c, le, 0, 2, 2, DWARF)
+HANDLE_DW_OP(0x2d, lt, 0, 2, 2, DWARF)
+HANDLE_DW_OP(0x2e, ne, 0, 2, 2, DWARF)
+HANDLE_DW_OP(0x2f, skip, 1, -1, 2, DWARF)
```
Comment (adrian-prantl) [inline llvm/include/llvm/BinaryFormat/Dwarf.def:738]: Like `DW_OP_bra`, this should have arity 0.

File: llvm/include/llvm/BinaryFormat/Dwarf.def [hunk 2]
```diff
@@ -51,7 +51,7 @@
 #endif
 
 #ifndef HANDLE_DW_OP
-#define HANDLE_DW_OP(ID, NAME, VERSION, VENDOR)
+#define HANDLE_DW_OP(ID, NAME, OPERANDS, ARITY, VERSION, VENDOR)
```
Comment (adrian-prantl) [inline llvm/include/llvm/BinaryFormat/Dwarf.def:54]: Is the number of operands ever useful without knowing the encoding of the operands?

File: llvm/include/llvm/BinaryFormat/Dwarf.def [hunk 3]
```diff
@@ -694,197 +694,197 @@ HANDLE_DW_FORM(0x1f21, GNU_strp_alt, 0, GNU)
...
 // New in DWARF v3:
-HANDLE_DW_OP(0x97, push_object_address, 3, DWARF)
-HANDLE_DW_OP(0x98, call2, 3, DWARF)
-HANDLE_DW_OP(0x99, call4, 3, DWARF)
-HANDLE_DW_OP(0x9a, call_ref, 3, DWARF)
-HANDLE_DW_OP(0x9b, form_tls_address, 3, DWARF)
-HANDLE_DW_OP(0x9c, call_frame_cfa, 3, DWARF)
-HANDLE_DW_OP(0x9d, bit_piece, 3, DWARF)
+HANDLE_DW_OP(0x97, push_object_address, 0, 0, 3, DWARF)
+HANDLE_DW_OP(0x98, call2, 1, 1, 3, DWARF)
```
Comment (adrian-prantl) [inline llvm/include/llvm/BinaryFormat/Dwarf.def:844]: I think these should have arity 0?

File: llvm/include/llvm/BinaryFormat/Dwarf.def [hunk 4]
```diff
@@ -694,197 +694,197 @@ HANDLE_DW_FORM(0x1f21, GNU_strp_alt, 0, GNU)
...
+HANDLE_DW_OP(0x99, call4, 1, 1, 3, DWARF)
+HANDLE_DW_OP(0x9a, call_ref, 1, 1, 3, DWARF)
+HANDLE_DW_OP(0x9b, form_tls_address, 0, 1, 3, DWARF)
+HANDLE_DW_OP(0x9c, call_frame_cfa, 0, 0, 3, DWARF)
+HANDLE_DW_OP(0x9d, bit_piece, 2, 1, 3, DWARF)
```
Comment (adrian-prantl) [inline llvm/include/llvm/BinaryFormat/Dwarf.def:849]: like DW_OP_piece this should have arity 0

File: llvm/include/llvm/BinaryFormat/Dwarf.def [hunk 5]
```diff
@@ -694,197 +694,197 @@ HANDLE_DW_FORM(0x1f21, GNU_strp_alt, 0, GNU)
...
 // New in DWARF v4:
-HANDLE_DW_OP(0x9e, implicit_value, 4, DWARF)
-HANDLE_DW_OP(0x9f, stack_value, 4, DWARF)
+HANDLE_DW_OP(0x9e, implicit_value, 2, 0, 4, DWARF)
+HANDLE_DW_OP(0x9f, stack_value, 0, 1, 4, DWARF)
```
Comment (adrian-prantl) [inline llvm/include/llvm/BinaryFormat/Dwarf.def:852]: I would use 0 here too (but it's a weaker requirement than in DW_OP_piece).

**Review by adrian-prantl — COMMENTED** [review]
This is a nice cleanup. I found a few minor issues in the table.
I'm also not sure the number of operands is ever useful without having the encoding ready, but maybe for DIExpression it actually is (because there we always use the same encoding).

**Comment by JDevlieghere** [comment]
> I'm also not sure the number of operands is ever useful without having the encoding ready, but maybe for DIExpression it actually is (because there we always use the same encoding).

Yeah, I came to the same conclusion, which is why LLDB isn't using the data. I primarily added it to avoid mistakes where someone confuses the number of operands with the arity.

File: llvm/include/llvm/BinaryFormat/Dwarf.def (hunk 3 above)
Comment (JDevlieghere) [reply llvm/include/llvm/BinaryFormat/Dwarf.def:844]: Yup, this is an example where I read "two operands" and confused that with "two stack entries".

**Review by clayborg — COMMENTED** [review]
Looks good to me as long as the data is correct. Cleans up the code nicely. 

It would be possible to encode the argument values using DW_FORM enumerations with variadic macros. If we move the `OPERANDS, ARITY` to the end and have ARITY come first:
```
#define HANDLE_DW_OP(ID, NAME, VERSION, VENDOR, ARITY, OPERANDS, ...)  
```
And then for DW_OP_addr:
```
HANDLE_DW_OP(0x03, addr, 2, DWARF, 0, 1, DW_FORM_addr)
```
Then clients can use the `__VA_ARGS__` if they want this info.

**Comment by JDevlieghere** [comment]
@clayborg that's a cool idea. We might be able to use that to implement `GetOpcodeDataSize`. That would definitely be a separate PR though.

**Review by adrian-prantl — APPROVED** [review]

**Review by clayborg — APPROVED** [review]

**Pull request merged**